stock_picking_invoicing/models/stock.py

```diff
@@ -0,0 +1,458 @@
+        
+        
+    @api.multi
+    def _get_price_unit_invoice(self, inv_type):
+        """ Gets price unit for invoice
+        @param move_line: Stock move lines
+        @param type: Type of invoice
+        @return: The price unit for the move line
+        """
+        
+        if type in ('in_invoice', 'in_refund'):
+            return move_line.product_id.price        
+        else:
+            # If partner given, search price in its sale pricelist
+            if self.partner_id and self.partner_id.property_product_pricelist:
+                self = self.with_context(
+                                         partner=self.partner_id.id,
+                                         quantity=self.product_uom_qty,
+                                         date=self.date,
+                                         pricelist=self.partner_id.property_product_pricelist.id,
+                                         uom=self.product_uom.id
+                                            )
+        result = self.product_id.price        
+
+        return result
+    
+    @api.multi
+    def _get_moves_taxes(self, moves, inv_type):
+        
+        #extra moves with the same picking_id and product_id of a move have the same taxes
+        extra_move_tax = {}
+        is_extra_move = {}
+        for move in moves:
+            if move.picking_id:
+                is_extra_move[move.id] = True
+                if not (move.picking_id, move.product_id) in extra_move_tax:
+                    extra_move_tax[move.picking_id, move.product_id] = 0
+            else:
+                is_extra_move[move.id] = False
+        return (is_extra_move, extra_move_tax)
+    
+
+#----------------------------------------------------------
+# Picking
+#----------------------------------------------------------
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    #===========================
+    # Views
+    #===========================
+    
+    @api.multi
+    def set_to_be_invoiced(self):
+        for picking in self:
+            if picking.invoice_state == '2binvoiced':
+                raise Warning(_("Can't update invoice control for picking %s: "
+                              "It's 'to be invoiced' yet") % picking.name)
+            if picking.invoice_state in ('none', 'invoiced'):
+                if picking.invoice_id:
+                    raise Warning(_('Picking %s has linked invoice %s') %
+                                  (picking.name, picking.invoice_id.number))
+            picking.invoice_state = '2binvoiced'
+            for move in picking.move_lines:
+                if move.invoice_state != 'invoiced':
+                    move.invoice_state = '2binvoiced'
+        return True
+    
+    @api.multi
+    def set_invoiced(self):
+        for picking in self:
+            if picking.invoice_state == 'invoiced' or picking.invoice_id :
+                raise Warning(_("Can't update invoice control for picking %s: "
```

Comment:
Invoice state cannot be updated for picking %s as it is already set as "Invoiced"